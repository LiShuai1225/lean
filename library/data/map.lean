/-
Copyright (c) 2016 Microsoft Corporation. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Author: Leonardo de Moura
-/
definition map (A B : Type) := A → option B

namespace map
variables {A B : Type}
open tactic

definition empty : map A B :=
λ a, none

definition lookup (k : A) (m : map A B) : option B :=
m k

definition select [inhabited B] (k : A) (m : map A B) : B :=
match lookup k m with
| (some v) := v
| none := default B
end

theorem ext (m₁ m₂ : map A B) : (∀ a, lookup a m₁ = lookup a m₂) → m₁ = m₂ :=
funext

theorem lookup_empty (k : A) : lookup k (empty : map A B) = none :=
rfl

variable [decidable_eq A]

definition insert (k : A) (v : B) (m : map A B) : map A B :=
λ a, if a = k then some v else m a

theorem lookup_insert (k : A) (v : B) (m : map A B) : lookup k (insert k v m) = some v :=
by unfold ["map" <.> "insert", "map" <.> "lookup"] >> dsimp >> rewrite "if_pos" >> reflexivity

theorem lookup_insert_of_ne {k₁ k₂ : A} (v : B) (m : map A B) : k₁ ≠ k₂ → lookup k₁ (insert k₂ v m) = lookup k₁ m :=
by intros >> unfold ["map" <.> "insert", "map" <.> "lookup"] >> dsimp >> rewrite "if_neg" >> assumption

definition erase (k : A) (m : map A B) : map A B :=
λ a, if a = k then none else m a

theorem lookup_erase (k : A) (v : B) (m : map A B) : lookup k (erase k m) = none :=
by unfold ["map" <.> "erase", "map" <.> "lookup"] >> dsimp >> rewrite "if_pos" >> reflexivity

theorem lookup_erase_of_ne {k₁ k₂ : A} (v : B) (m : map A B) : k₁ ≠ k₂ → lookup k₁ (erase k₂ m) = lookup k₁ m :=
by intros >> unfold ["map" <.> "erase", "map" <.> "lookup"] >> dsimp >> rewrite "if_neg" >> assumption

theorem erase_empty (k : A) : erase k empty = (empty : map A B) :=
funext (λ a, by unfold ["map" <.> "erase", "map" <.> "empty"] >> rewrite "if_t_t")

theorem erase_insert {k : A} {m : map A B} (v : B) : lookup k m = none → erase k (insert k v m) = m :=
assume h, funext (λ a, decidable.by_cases
  (suppose a = k, by get_local "this" >>= subst >> unfold ["map" <.> "erase", "map" <.> "insert"] >> rewrite "if_pos" >> symmetry >> assumption >> reflexivity)
  (suppose a ≠ k, by unfold ["map" <.> "erase", "map" <.> "insert"] >> rewrite "if_neg" >> dsimp >> rewrite "if_neg" >> repeat assumption))

end map
