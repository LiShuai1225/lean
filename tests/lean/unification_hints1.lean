import data.list data.nat
open list nat

structure unification_constraint := {A : Type} (lhs : A) (rhs : A)
structure unification_hint := (pattern : unification_constraint) (constraints : list unification_constraint)

namespace toy
constants (A : Type.{1}) (f h : A → A) (x y z : A)
definition g [irreducible] (x y : A) : A := f z

#unify (g x y), (f z)

definition toy_hint [unify] (x y : A) : unification_hint :=
  unification_hint.mk (unification_constraint.mk (g x y) (f z)) []

#unify (g x y), (f z)
print [unify]

end toy

namespace add
constants (n : ℕ)

#unify (n + 1), succ n

definition add_zero_hint [unify] (m n : ℕ) [has_add ℕ] [has_one ℕ] [has_zero ℕ] : unification_hint :=
  unification_hint.mk (unification_constraint.mk (m + 1) (succ n)) [unification_constraint.mk m n]

#unify (n + 1), (succ n)
print [unify]

end add

namespace canonical
structure Canonical := (carrier : Type) (op : carrier → carrier)
attribute Canonical.carrier [irreducible]

constants (A : Type.{1}) (f : A → A) (x : A)
definition A_canonical : Canonical := Canonical.mk A f

#unify (Canonical.carrier A_canonical), A

definition Canonical_hint [unify] (C : Canonical) : unification_hint :=
  unification_hint.mk (unification_constraint.mk (Canonical.carrier C) A) [unification_constraint.mk C A_canonical]

-- TODO(dhs): we mark carrier as irreducible and prove A_canonical explicitly to work around the fact that
-- the default_type_context does not recognize the elaborator metavariables as metavariables,
-- and so cannot perform the assignment.
#unify (Canonical.carrier A_canonical), A
print [unify]

end canonical

print [unify] canonical
