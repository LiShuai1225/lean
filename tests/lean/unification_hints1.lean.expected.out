fail
success
unification hints:
(toy.g, toy.f) g #1 #0 =?= f z {}
fail
success
unification hints:
(add, nat.succ) #4 + 1 =?= succ #3 {#4 =?= #3, }
fail
success
unification hints:
(canonical.Canonical.carrier, canonical.A) Canonical.carrier #0 =?= A {#0 =?= A_canonical, }
unification hints at namespace 'canonical':
(canonical.Canonical.carrier, canonical.A) canonical.Canonical.carrier #0 =?=
  canonical.A {#0 =?= canonical.A_canonical, }
